pkg/shardddl/optimism/lock_test.go

```diff
@@ -1532,6 +1540,123 @@ func (t *testLock) TestAddDifferentFieldLenColumns(c *C) {
+				db: {tbls[0]: 0, tbls[1]: 0},
+			},
+		}
+
+		colm1 = map[string]map[string]map[string]map[string]map[string]interface{}{
+			ID: {
+				"b": {source: {db: {tbls[0]: struct{}{}}}},
+				"c": {source: {db: {tbls[0]: struct{}{}}}},
+			},
+		}
+		colm2 = map[string]map[string]map[string]map[string]map[string]interface{}{
+			ID: {
+				"c": {source: {db: {tbls[0]: struct{}{}}}},
+			},
+		}
+	)
+	col, err := GetColumnName(ID, DDLs1[0], ast.AlterTableDropColumn)
+	c.Assert(col, Equals, "c")
+	c.Assert(err, IsNil)
+	col, err = GetColumnName(ID, DDLs2[0], ast.AlterTableDropColumn)
+	c.Assert(col, Equals, "b")
+	c.Assert(err, IsNil)
+
+	// the initial status is synced but not resolved.
+	t.checkLockSynced(c, l)
+	t.checkLockNoDone(c, l)
+	c.Assert(l.IsResolved(), IsFalse)
+
+	// TrySync for the first table, drop column c
+	DDLs, err := l.TrySync(newInfoWithVersion(task, source, db, tbls[0], downSchema, downTable, DDLs1, ti0, []*model.TableInfo{ti1}, vers), tts)
+	c.Assert(err, IsNil)
+	c.Assert(DDLs, DeepEquals, []string{})
+	c.Assert(l.versions, DeepEquals, vers)
+	c.Assert(l.IsResolved(), IsFalse)
+
+	// TrySync for the first table, drop column b
+	DDLs, err = l.TrySync(newInfoWithVersion(task, source, db, tbls[0], downSchema, downTable, DDLs2, ti1, []*model.TableInfo{ti2}, vers), tts)
+	c.Assert(err, IsNil)
+	c.Assert(DDLs, DeepEquals, []string{})
+	c.Assert(l.versions, DeepEquals, vers)
+	c.Assert(l.IsResolved(), IsFalse)
+
+	colm, _, err := GetAllDroppedColumns(etcdTestCli)
+	c.Assert(err, IsNil)
+	c.Assert(colm, DeepEquals, colm1)
+
+	// TrySync for the second table, drop column b, this column should be dropped
+	DDLs, err = l.TrySync(newInfoWithVersion(task, source, db, tbls[1], downSchema, downTable, DDLs2, ti0, []*model.TableInfo{ti3}, vers), tts)
+	c.Assert(err, IsNil)
+	c.Assert(DDLs, DeepEquals, DDLs2)
+	c.Assert(l.versions, DeepEquals, vers)
+	c.Assert(l.IsResolved(), IsFalse)
+	// Simulate watch done operation from dm-worker
+	c.Assert(l.DeleteColumnsByDDLs(DDLs), IsNil)
+
+	colm, _, err = GetAllDroppedColumns(etcdTestCli)
+	c.Assert(err, IsNil)
+	c.Assert(colm, DeepEquals, colm2)
+
+	// TrySync for the first table, add column b, should succeed, because this column is fully dropped in the downstream
+	DDLs, err = l.TrySync(newInfoWithVersion(task, source, db, tbls[0], downSchema, downTable, DDLs3, ti2, []*model.TableInfo{ti1}, vers), tts)
+	c.Assert(err, IsNil)
+	c.Assert(DDLs, DeepEquals, DDLs3)
+	c.Assert(l.versions, DeepEquals, vers)
+	c.Assert(l.IsResolved(), IsFalse)
+
+	// TrySync for the first table, add column b, should fail, because this column isn't fully dropped in the downstream
+	_, err = l.TrySync(newInfoWithVersion(task, source, db, tbls[0], downSchema, downTable, DDLs4, ti1, []*model.TableInfo{ti0}, vers), tts)
+	c.Assert(err, ErrorMatches, ".*add column c that wasn't fully dropped in downstream.*")
+	c.Assert(l.IsResolved(), IsFalse)
+
+	// TrySync for the second table, drop column b, this column should be dropped
```

Comment:
repeated? we didn't add column b after line 1618 🤔

Reply:
https://github.com/pingcap/dm/pull/1510/commits/fb5857aef167f500d7f5250f36cf77b11e3d889c